cpp/include/cugraph/graph_functions.hpp

```diff
@@ -679,42 +681,43 @@ extract_induced_subgraphs(
  * compute_gpu_id_from_ext_edge_endpoints_t to every edge should return the local GPU ID for this
  * function to work (edges should be pre-shuffled).
  * @param edgelist_dsts Vector of edge destination vertex IDs.
- * @param edgelist_weights Vector of edge weights.
- * @param edgelist_id_type_pairs Vector of edge ID and type pairs.
+ * @param edgelist_property1 Vector of property1 values for edges
+ * @param edgelist_property2 Vector of property2 values for edges
+ * @param edgelist_property3 Vector of property3 values for edges
  * @param graph_properties Properties of the graph represented by the input (optional vertex list
  * and) edge list.
  * @param renumber Flag indicating whether to renumber vertices or not (must be true if @p multi_gpu
  * is true).
  * @param do_expensive_check A flag to run expensive checks for input arguments (if set to `true`).
- * @return Tuple of the generated graph and optional edge_property_t objects storing edge weights
- * and edge IDs & types (valid if @p edgelist_weights.has_value() and @p
- * edgelist_id_type_pairss.has_value() are true, respectively) and a renumber map (if @p renumber is
- * true).
+ * @return Tuple of the generated graph and optional edge_property_t objects storing the provided
+ * edge properties and a renumber map (if @p renumber is true).
  */
 template <typename vertex_t,
           typename edge_t,
-          typename weight_t,
-          typename edge_type_t,
+          typename property1_t,
+          typename property2_t,
+          typename property3_t,
```

Comment:
Longer term I wonder if the solution is to type erase the edge properties, since they are infrequently used, and dispatch at runtime to the proper type.

I had moved to this solely to offer flexibility if we decide a different combination of up to 3 properties was desirable for some algorithm.  You are correct we would need to instantiate whatever combinations we need up front (unless we type erase).

Reply:
Yeah, I agree that type erasing might be a good longer term solution, but we are not there yet (and we're not processing three arrays separately but building a tuple out of these three types and passing the tuple to a functor, so we actually still need to explicitly instantiate in the lower level if I am not mistaken).

If we're not planning to move to the type erasing direction in short term, I am a bit concerned about the inconsistency this will bring to our API.

Reply:
We can't go to type erasing until we've built up some infrastructure for that inside the C++ code base.  I wasn't think we'd try and go there yet.  We haven't even really talked about putting that on our road map yet.

I guess I was trying to be forward thinking about other potential uses.  Today the only properties we support are weight, edge id, edge type, where weight is of type `weight_t`, edge id is of type `edge_t` and edge type is of type `int32_t`.  This more generic naming would allow us to support, for example, creating a graph with multiple weights and picking which one to apply to different algorithms (a feature supported by networkx).  We do nothing inside this function, or inside any of our algorithms, that really uses anything but a weight.  The sampling algorithm blindly returns whatever values are in the properties.  It doesn't matter that they are an edge id and an edge type.

I'm OK changing them back if we think that's better until we have a type erased solution.

Reply:
I have thought more about this, and there are few things to consider.

1) Can users currently use this without knowing that property1_t is weight_t, property2_t is edge_t, and property3_t is edge_type_t? They can't if I am not mistaken. I guess property?_t makes this API more difficult to use.
2) Should we really go to the type erasing path and what do we really get? We need to explicitly instantiate in lower level if we're willing to pass a thrust::tuple to a functor. Otherwise, individual functors need to deal with type erased variables, and this adds overhead in both coding complexity and performance.
3) If we're passing many properties in a type erased fashion, we may better take a different approach. If we create an array per property and if we access 100 properties, this will involve 100 4-8 byte random accesses. If we store 100 properties in a packed fashion, this will involve 1 access of size 100 * (4-8) bytes. In this case, we may store device_span to the packed data in a view object and better pass this device_span object to individual functors. Assuming that individual functors know how to unpack the array, the API does not explicitly specify the types.

I guess using property?_t is little pre-mature at this point as we are not sure this choice will really be future proof or not at this point.

Reply:
I agree this makes the API confusing, since there are algorithms that use weight, for instance, and it isn't clear whether the weight is property1, property2, or property3.  If users could name the properties and weren't restricted in the types of each property, then the algorithms would just require the user specify which property to treat as weight.  But I wouldn't do that without type erasure.  I concur with Seungwha on leaving them as they are.

Reply:
I will revert this.

Type erasing is, admittedly, a big design conversation we should have independent of this activity.  If we go the type erasing route it really needs to be for more than just this special case, and it would likely need to have effects that ripple through all of the primitives to truly be effective.  Cudf uses the type erased route all the way to cuda kernel calls and has had good success with it.  But we would need to evaluate and decide to go that route (if it makes sense) and lay out a comprehensive plan, not do it piecemeal.